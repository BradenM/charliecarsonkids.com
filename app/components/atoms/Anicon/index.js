/**
 *
 * Anicon
 *
 * Atom
 */

import { Lottie } from '@crello/react-lottie';
import PropTypes from 'prop-types';
import React, { useRef, useState } from 'react';
import styled from 'styled-components';
import { AniconURL } from 'utils/getStatic';

const HoverContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
`;

const Anicon = ({ name, loop, rotate, hoverable, moveTo, ...props }) => {
  const [direction, setDirection] = useState(-1);

  const animRef = useRef(null);

  const handleMouseEvent = () => {
    const {
      current: { animationItem },
    } = animRef;
    animationItem.play();
    setDirection(direction > 0 ? -1 : 1);
  };

  const handleCatchFrame = () => {
    const {
      current: { animationItem },
    } = animRef;
    const { currentFrame } = animationItem;
    const delim = moveTo - 0.5;
    if (currentFrame >= delim) {
      animationItem.pause();
    }
  };

  const renderConfig = {
    autoplay: false,
    path: AniconURL(name),
    loop: false,
  };

  return (
    <HoverContainer
      className="anicon-container"
      onMouseEnter={hoverable ? handleMouseEvent : () => {}}
      onMouseLeave={hoverable ? handleMouseEvent : () => {}}
    >
      <Lottie
        direction={direction}
        ref={animRef}
        lottieEventListeners={[
          { name: 'enterFrame', callback: handleCatchFrame },
          {
            name: 'data_ready',
            callback: () => animRef.current.animationItem.stop(),
          },
        ]}
        style={{
          transform: `rotate(${rotate}deg)`,
          cursor: 'pointer',
        }}
        config={renderConfig}
        {...props}
      />
    </HoverContainer>
  );
};

Anicon.defaultProps = {
  rotate: '0',
  loop: false,
  hoverable: true,
};

Anicon.propTypes = {
  name: PropTypes.string.isRequired,
  rotate: PropTypes.string,
  loop: PropTypes.bool,
  hoverable: PropTypes.bool,
  ...Lottie.propTypes,
};

export default Anicon;
