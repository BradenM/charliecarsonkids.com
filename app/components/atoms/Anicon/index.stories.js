/**
 *
 * Stories for Anicon
 *
 * Atom
 */

import { withActions } from '@storybook/addon-actions';
import { boolean, number, text } from '@storybook/addon-knobs';
import React from 'react';
import Anicon from '.';

export default {
  title: 'Design|Atoms/Anicon',
  component: Anicon,
  decorators: [withActions('mouseenter', 'mouseleave')],
};

// const eventsWithName = actions('onMouseEnter', 'onMouseLeave');

export const searchAnicon = () => (
  <Anicon name="search" width="100px" moveTo={14} />
);

export const withDynamicVariables = () => {
  const knobs = {
    name: text('Icon', 'shopping'),
    width: '100px',
    speed: number('Speed', 1),
    rotate: text('Rotate', '0'),
    hoverable: boolean('Hoverable', true),
    moveTo: number('Frame on Hover', 14),
  };

  return <Anicon {...knobs} />;
};
