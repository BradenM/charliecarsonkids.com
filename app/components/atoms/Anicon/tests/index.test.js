/**
 *
 * Tests for Anicon
 *
 *
 */

import { render } from '@testing-library/react';
import React from 'react';
import Anicon from '../index';

const renderComponent = props => render(<Anicon name="icon" {...props} />);

describe('<Anicon />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<Anicon name="icon" />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = renderComponent();
    expect(firstChild).toMatchSnapshot();
  });
});
