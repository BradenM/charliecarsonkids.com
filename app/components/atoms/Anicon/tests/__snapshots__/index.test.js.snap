// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Anicon /> Should render and match the snapshot 1`] = `
<div
  class="Anicon__HoverContainer-sc-119yqeq-0 iCLpQU anicon-container"
>
  <div
    style="width: 100%; height: 100%; transform: rotate(0deg); cursor: pointer;"
  />
</div>
`;
